golang/cosmos/x/swingset/keeper/msg_server.go

```diff
@@ -12,23 +13,73 @@ type msgServer struct {
 	Keeper
 }
 
-// NewMsgServerImpl returns an implementation of the bank MsgServer interface
+// NewMsgServerImpl returns an implementation of the swingset MsgServer interface
 // for the provided Keeper.
 func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 	return &msgServer{Keeper: keeper}
 }
 
 var _ types.MsgServer = msgServer{}
 
+type ActionMeta struct {
+	Type        string `json:"type"`
+	BlockHeight int64  `json:"blockHeight"`
+	BlockTime   int64  `json:"blockTime"`
+}
 type deliverInboundAction struct {
-	Type        string          `json:"type"`
-	Peer        string          `json:"peer"`
-	Messages    [][]interface{} `json:"messages"`
-	Ack         uint64          `json:"ack"`
-	BlockHeight int64           `json:"blockHeight"`
-	BlockTime   int64           `json:"blockTime"`
+	*ActionMeta
+	Peer     string          `json:"peer"`
+	Messages [][]interface{} `json:"messages"`
+	Ack      uint64          `json:"ack"`
+}
+type walletAction struct {
+	*ActionMeta
+	Owner  string `json:"owner"`
+	Action string `json:"action"`
+}
+type walletSpendAction struct {
+	*ActionMeta
+	Owner       string `json:"owner"`
+	SpendAction string `json:"spendAction"`
+}
+type provisionAction struct {
+	*ActionMeta
+	*types.MsgProvision
+	AutoProvision bool `json:"autoProvision"`
+}
+type installBundleAction struct {
+	*ActionMeta
+	*types.MsgInstallBundle
+}
+
+// finishAction adds to an "action" instance meta-information that is necessary
+// for it to be accepted by the swingset kernel (most notably a type string).
+func finishAction[T *deliverInboundAction | *walletAction | *walletSpendAction | *provisionAction | *installBundleAction](action T, ctx sdk.Context) T {
```

Comment:
Please put context first, following the existing convention of other contextual functions:
```suggestion
func finishAction[T *deliverInboundAction | *walletAction | *walletSpendAction | *provisionAction | *installBundleAction](ctx sdk.Context, action T) T {
```